swri_roscpp/include/swri_roscpp/topic_service_client.h

```diff
@@ -0,0 +1,173 @@
+// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+// ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// *****************************************************************************
+#ifndef SWRI_ROSCPP_TOPIC_SERVICE_CLIENT_H_
+#define SWRI_ROSCPP_TOPIC_SERVICE_CLIENT_H_
+
+#include <ros/node_handle.h>
+
+#include <boost/thread/mutex.hpp>
+
+#include <map>
+
+namespace swri
+{
+
+template<class MReq, class MRes>
+class TopicServiceClientRaw
+{
+ private:
+  typedef
+  boost::mutex request_lock_;
+  ros::Subscriber response_sub_;
+  ros::Publisher request_pub_;
+  boost::shared_ptr<MRes> response_;
+
+  ros::Duration timeout_;
+  std::string name_;
+  std::string service_name_;
+
+  int sequence_;
+
+ public:
+  TopicServiceClientRaw() : sequence_(0), timeout_(ros::Duration(4.0))
+  {
+
+  }
+
+  void initialize(ros::NodeHandle &nh,
+                const std::string &service,
+                const std::string &client_name = "")
+  {
+    name_ = client_name.length() ? client_name : ros::this_node::getName();
+    std::string rservice = nh.resolveName(service);
+    service_name_ = rservice;
+    response_sub_ = nh.subscribe(rservice + "/response", 10, &TopicServiceClientRaw<MReq, MRes>::response_callback, this);
+    request_pub_ = nh.advertise<MReq>(rservice + "/request", 10);
+  }
+
+  bool call(MReq& request, MRes& response)
+  {
+    boost::mutex::scoped_lock scoped_lock(request_lock_);
+
+    // block for response
+    request.srv_header.stamp = ros::Time::now();
+    request.srv_header.sequence = sequence_;
+    request.srv_header.sender = name_;
+
+    // Wait until we get a subscriber and publisher
+    while (request_pub_.getNumSubscribers() == 0 && response_sub_.getNumPublishers() == 0)
+    {
+      ros::Duration(0.002).sleep();
+      ros::spinOnce();
+
+      if (ros::Time::now() - request.srv_header.stamp > timeout_)
+      {
+        return false;
+      }
+    }
+    request_pub_.publish(request);
+
+    // Wait until we get a response
+    while (!response_ && ros::Time::now() - request.srv_header.stamp < timeout_)
+    {
+      ros::Duration(0.002).sleep();
+      ros::spinOnce();
+    }
+
+    sequence_++;
+    if (response_)
+    {
+      response = *response_;
+      response_.reset();
+      return response.srv_header.result;
+    }
+    return false;
+  }
+
+  std::string getService()
+  {
+    return service_name_;
+  }
+
+  bool exists()
+  {
+    return request_pub_.getNumSubscribers() > 0;
+  }
+
+  //bool call_async(MReq& request, boost::function<void(const MReq&, const MRes&)>& fun);
```

Comment:
What's this here for?

Reply:
Was a potential reminder. Will just remove